### PR TITLE
Raise exception when admin group name is not found

### LINKED ISSUE #2187: No clear error message when admin group name is incorrect in context
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **deploying** a Data Safe Haven.
- [x] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Operating System:** macOs
- **Data Safe Haven version:** develop

## :package: Packages

<details>
<summary>List of packages</summary>

```none
Paste list of packages here
```

</details>

## :no_entry_sign: Describe the problem

Deploying an SHM fails if the `admin-group-name` does not correspond to a known security group. This situation is already captured in the codebase, but does not return an error. It'd be much easier for a user to diagnose this problem if an appropriate error message were returned.

## :deciduous_tree: Log messages

<details>
<summary>Relevant log messages</summary>

```none
You are logged into the Azure CLI as:                                                                                                      
        user: Matthew Craddock (1b2a59f2-f3ea-42e7-b8b8-aeea30572764)                                                                      
        tenant: turing.ac.uk (4395f4a7-e455-4f95-8a9f-1fbaef6384f9)                                                                        
Are these details correct? [y/n] (y): y
Preparing to deploy testadmingroup SHM.                                                                                                    
Ensured that resource group shm-checkadmin-rg exists in uksouth.                                                                           
Ensured that managed identity shm-checkadmin-identity-reader exists.                                                                       
Ensured that storage account shmcheckadmin exists.                                                                                         
Ensured that storage container config exists.                                                                                              
Ensured that storage container pulumi exists.                                                                                              
Failed to create key vault shm-checkadmin-kv.                                                                                              
Failed to deploy resources needed by Pulumi.                                                                                               
Failed to deploy Data Safe Haven infrastructure. 
```

</details>

## :recycle: To reproduce

Deploy an SHM after setting up a context with an incorrect `admin-group-name`

### PR DESCRIPTION
**Pull request by craddm**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :arrow_heading_up: Summary

Raises an exception when the admin group name is specified incorrectly in the context. 

### :closed_umbrella: Related issues

Closes #2187 

### :microscope: Tests

Tested locally

### CODE REVIEW

File: data_safe_haven/config/shm_config.py [hunk 1]
```diff
@@ -27,10 +28,14 @@ def from_args(
     ) -> SHMConfig:
         """Construct an SHMConfig from arguments."""
         azure_sdk = AzureSdk(subscription_name=context.subscription_name)
-        admin_group_id = (
-            azure_sdk.entra_directory.get_id_from_groupname(context.admin_group_name)
-            or "admin-group-id-not-found"
-        )
```
Comment (JimMadge) [inline data_safe_haven/config/shm_config.py:33]: Do we want to continue here if the id cannot be found?

Comment (JimMadge) [reply data_safe_haven/config/shm_config.py:33]: Looks like this change might have been added to satisfy mypy as the function can return `None` https://github.com/alan-turing-institute/data-safe-haven/commit/b5e08768fc2cca8d52ff84b728c36c21a54b5792

**Comment by JimMadge** [comment]
Returning `None` is important [here](https://github.com/alan-turing-institute/data-safe-haven/blob/cdd76a3b51a8278ab1562f90ab8637ad8ab5276b/data_safe_haven/external/api/graph_api.py#L317) where if there is no group then one is created.

**Comment by JimMadge** [comment]
Also used in `remove_user_from_group` and `add_user_to_group` where `None` is not handled.

A good solution could be adding a method like

```python
def ensure_id_from_groupname(name):
   if (id := get_id_from_groupname(name)):
       return id
   else:
       raise Exception
```

File: data_safe_haven/config/shm_config.py [hunk 2]
```diff
@@ -27,10 +28,14 @@ def from_args(
...
+        try:
+            admin_group_id = azure_sdk.entra_directory.get_id_from_groupname(
+                context.admin_group_name
+            )
+        except DataSafeHavenMicrosoftGraphError as exc:
+            msg = f"Failed to get ID for group '{context.admin_group_name}'"
+            raise DataSafeHavenMicrosoftGraphError(msg) from exc
+
```
Comment (jemrobinson) [inline data_safe_haven/config/shm_config.py]: How about this instead of adding a new method?

```suggestion
        admin_group_id := azure_sdk.entra_directory.get_id_from_groupname(
            context.admin_group_name
        )
        if not admin_group_id:
            msg = f"Failed to get ID for group '{context.admin_group_name}'"
            raise DataSafeHavenMicrosoftGraphError(msg) from exc

```

Then we can leave the function as returning `str | None` and avoid needing to add exception handling. Note that this will also throw the exception if the ID of the group is `""` but I think that's reasonable, since a valid ID will be a GUID.

File: data_safe_haven/external/api/graph_api.py [hunk 3]
```diff
@@ -515,7 +515,7 @@ def get_service_principal_by_name(
         except (DataSafeHavenMicrosoftGraphError, StopIteration):
             return None
 
-    def get_id_from_groupname(self, group_name: str) -> str | None:
+    def get_id_from_groupname(self, group_name: str) -> str:
```
Comment (jemrobinson) [inline data_safe_haven/external/api/graph_api.py]: If taking my suggestion above, we should revert this change.

File: data_safe_haven/config/shm_config.py (hunk 2 above)
Comment (JimMadge) [reply data_safe_haven/config/shm_config.py]: Won't this raise the exception if `admin_group_id` is `None`? `not None is True`

**Comment by craddm** [comment]
> Also used in `remove_user_from_group` and `add_user_to_group` where `None` is not handled.
> 
> A good solution could be adding a method like
> 
> ```python
> def ensure_id_from_groupname(name):
>    if (id := get_id_from_groupname(name)):
>        return id
>    else:
>        raise Exception
> ```

This feels weird to me.

We have a function called `get_id_from_groupname` which gets the id if the group exists or returns `None`, and now a function called `ensure_id_from_groupname` which also gets the id if the group exists, but throws an exception if it doesn't exist. I feel like that is a bit confusing, especially given that (almost) all the other `ensure_*` functions check if something exists and create it if it doesn't.

It feels like we should have a function that checks if the group exists, and then one that returns the group id (but is only used when we know the group exists). The slight problem with that is the way to check if the group exists is to try to retrieve the whole list of groups and search through it, which you then would have to do multiple times instead of once.

**Comment by JimMadge** [comment]
This is the internal API so I wouldn't worry _too_ much about names.

You could change how the current `create_group` works to return the name if it exists but that might be too much for one function.

Making two API calls when you only need one feels wasteful.
I still think just adding a method for the common pattern of ensuring the groups exists makes most sense. It is just taking the code that would be duplicated in a number of places and instead writing it once.

**Comment by craddm** [comment]
> This is the internal API so I wouldn't worry _too_ much about names.
> 
> You could change how the current `create_group` works to return the name if it exists but that might be too much for one function.
> 
> Making two API calls when you only need one feels wasteful. I still think just adding a method for the common pattern of ensuring the groups exists makes most sense. It is just taking the code that would be duplicated in a number of places and instead writing it once.

Agree about the multiple API calls, which would be easy enough to stop by instead passing the list of groups to the functions from a separate call to `read_groups()`. But anyway, gone back to your suggested style

File: data_safe_haven/config/shm_config.py (hunk 2 above)
Comment (jemrobinson) [reply data_safe_haven/config/shm_config.py]: Yes, if it's `None` (the lookup failed) or `""` (the lookup returned empty string, which shouldn't be possible). That's what we want, right?

**Review by JimMadge — APPROVED** [review]
LGTM :+1:
